vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -1,1011 +1,527 @@
-                    "mor": obj,
-                    "hostname": hostname,
-                    "tags": [t for t in tags + instance_tags if t.split(":", 1)[0] not in self.excluded_host_tags],
-                }
-                if self.excluded_host_tags:
-                    mor["excluded_host_tags"] = [
-                        t for t in tags + instance_tags if t.split(":", 1)[0] in self.excluded_host_tags
-                    ]
-                obj_list[vimtype].append(mor)
-
-        self.log.debug("All objects with attributes cached in %s seconds.", time.time() - start)
-        return obj_list
-
-    @staticmethod
-    def _is_excluded(obj, properties, regexes, include_only_marked):
-        """
-        Return `True` if the given host or virtual machine is excluded by the user configuration,
-        i.e. violates any of the following rules:
-        * Do not match the corresponding `*_include_only` regular expressions
-        * Is "non-labeled" while `include_only_marked` is enabled (virtual machine only)
-        """
-        # Host
-        if isinstance(obj, vim.HostSystem):
-            # Based on `host_include_only_regex`
-            if regexes and regexes.get('host_include') is not None:
-                match = re.search(regexes['host_include'], properties.get("name", ""), re.IGNORECASE)
-                if not match:
-                    return True
-
-        # VirtualMachine
-        elif isinstance(obj, vim.VirtualMachine):
-            # Based on `vm_include_only_regex`
-            if regexes and regexes.get('vm_include') is not None:
-                match = re.search(regexes['vm_include'], properties.get("name", ""), re.IGNORECASE)
-                if not match:
-                    return True
-
-            # Based on `include_only_marked`
-            if include_only_marked:
-                monitored = False
-                for field in properties.get("customValue", ""):
-                    if field.value == VM_MONITORING_FLAG:
-                        monitored = True
-                        break  # we shall monitor
-                if not monitored:
-                    return True
-
-        return False
-
-    def _cache_morlist_raw(self, instance):
-        """
-        Fill the Mor objects queue that will be asynchronously processed later.
-        Resolve the vCenter `rootFolder` and initiate hosts and virtual machines
-        discovery.
-        """
-        i_key = self._instance_key(instance)
-        self.log.debug("Caching the morlist for vcenter instance %s", i_key)
+    def initiate_connection(self):
+        try:
+            self.log.info("Connecting to the vCenter API...")
+            self.api = VSphereAPI(self.instance)
+            self.log.info("Connected")
+        except APIConnectionError:
+            self.log.error("Cannot authenticate to vCenter API. The check will not run.")
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self.base_tags, hostname=None)
```

Comment:
`initiate_connection` would be only called once though so not very useful.

I don't think this is confusing, as there is no real difference between "Connecting to the API" and "Checking the health of the API" in that situation. A connection should be considered valid if and only if authentication is successful and the `CurrentTime` method can be run on the API.

I see only two situations where you could "connect" but not "check health":
- The service is healthy but the user you provided has no permission to run API queries
- The service is healthy but for some reason the API used is not compatible with the vCenter version.

In both situation, we would *not* be able to tell if the service is healthy or not so we can't submit a `vsphere.healthy` service check. Also if we split into two service checks, that mean we should submit `OK` for `vsphere.can_connect` because connection was successful. Yet this is clearly misleading.

I'm much happier with a single service check that tells whether or not the integration can communicate with the API.